### ISSUE TITLE
Console reverse search broken

### ISSUE THREAD

**Issue by maksay** [ipython/ipython#10014]
When trying to use reverse-search in `ipython console` / `jupyter console`, I am getting the following error:

```
I-search backward: Traceback (most recent call last):
  File "/Users/andriimaksai/anaconda/bin/jupyter-console", line 5, in <module>
    app.main()
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/jupyter_core/application.py", line 267, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/traitlets/config/application.py", line 596, in launch_instance
    app.start()
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/jupyter_console/app.py", line 151, in start
    self.shell.mainloop()
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/jupyter_console/ptshell.py", line 463, in mainloop
    self.interact()
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/jupyter_console/ptshell.py", line 448, in interact
    code = self.prompt_for_code()
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/jupyter_console/ptshell.py", line 408, in prompt_for_code
    reset_current_buffer=True)
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/interface.py", line 399, in run
    self._redraw()
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/interface.py", line 346, in _redraw
    self.renderer.render(self, self.layout, is_done=self.is_done)
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/renderer.py", line 429, in render
    extended_height=size.rows,
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/layout/containers.py", line 142, in write_to_screen
    sizes = self._divide_heigths(cli, write_position)
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/layout/containers.py", line 177, in _divide_heigths
    dimensions = [get_dimension_for_child(c, index) for index, c in enumerate(self.children)]
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/layout/containers.py", line 175, in get_dimension_for_child
    return c.preferred_height(cli, write_position.width, write_position.extended_height)
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/layout/containers.py", line 395, in preferred_height
    return self.content.preferred_height(cli, width, max_available_height)
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/layout/containers.py", line 128, in preferred_height
    dimensions = [c.preferred_height(cli, width, max_available_height) for c in self.children]
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/layout/containers.py", line 1000, in preferred_height
    cli, width - total_margin_width, max_available_height, wrap_lines),
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/layout/controls.py", line 502, in preferred_height
    content = self.create_content(cli, width, None)
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/layout/controls.py", line 614, in create_content
    ignore_case=ss.ignore_case))
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/buffer.py", line 1075, in document_for_search
    search_result = self._search(search_state, include_current_position=True)
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/buffer.py", line 1062, in _search
    result = search_once(working_index, document)
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/buffer.py", line 1052, in search_once
    document = Document(self._working_lines[i], len(self._working_lines[i]))
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/document.py", line 83, in __init__
    assert isinstance(text, six.text_type), 'Got %r' % text
AssertionError: Got 'ls'
```

Output of `python -c "import IPython; print(IPython.sys_info())"`

```
{'commit_hash': u'5c9c918',
 'commit_source': 'installation',
 'default_encoding': 'UTF-8',
 'ipython_path': '/Users/andriimaksai/anaconda/lib/python2.7/site-packages/IPython',
 'ipython_version': '5.1.0',
 'os_name': 'posix',
 'platform': 'Darwin-15.5.0-x86_64-i386-64bit',
 'sys_executable': '/Users/andriimaksai/anaconda/bin/python',
 'sys_platform': 'darwin',
 'sys_version': '2.7.12 |Anaconda custom (x86_64)| (default, Jul  2 2016, 17:43:17) \n[GCC 4.2.1 (Based on Apple Inc. build 5658) (LLVM build 2336.11.00)]'}
```

However, when I launch IPython from python, everything works normally: `python -c "import IPython; IPython.start_ipython()`. It also works as `python -m IPython`, but not as `python -m jupyter console`. Any ideas what might be the source of the problem?

**Comment by takluyver** [comment]
What version of prompt_toolkit do you have? If it's not the latest, update it. If it still happens, we'll ping Jonathan Slenders, who maintains prompt_toolkit. I'm starting to feel bad about pinging him a lot, which is why I'm deferring it until after you've checked that you've got the latest version.

**Comment by maksay** [comment]
Updated to the latest version of prompt_toolkit available in anaconda (1.0.7), basically the same error, just some line numbers changed (+here with ipython instead of jupyter, as in the previous example):

```
I-search backward: Traceback (most recent call last):
  File "/Users/andriimaksai/anaconda/bin/ipython", line 6, in <module>
    sys.exit(IPython.start_ipython())
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/IPython/__init__.py", line 119, in start_ipython
    return launch_new_instance(argv=argv, **kwargs)
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/traitlets/config/application.py", line 596, in launch_instance
    app.start()
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/IPython/terminal/ipapp.py", line 344, in start
    return self.subapp.start()
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/jupyter_console/app.py", line 151, in start
    self.shell.mainloop()
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/jupyter_console/ptshell.py", line 463, in mainloop
    self.interact()
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/jupyter_console/ptshell.py", line 448, in interact
    code = self.prompt_for_code()
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/jupyter_console/ptshell.py", line 408, in prompt_for_code
    reset_current_buffer=True)
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/interface.py", line 399, in run
    self._redraw()
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/interface.py", line 346, in _redraw
    self.renderer.render(self, self.layout, is_done=self.is_done)
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/renderer.py", line 436, in render
    extended_height=size.rows,
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/layout/containers.py", line 142, in write_to_screen
    sizes = self._divide_heigths(cli, write_position)
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/layout/containers.py", line 177, in _divide_heigths
    dimensions = [get_dimension_for_child(c, index) for index, c in enumerate(self.children)]
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/layout/containers.py", line 175, in get_dimension_for_child
    return c.preferred_height(cli, write_position.width, write_position.extended_height)
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/layout/containers.py", line 395, in preferred_height
    return self.content.preferred_height(cli, width, max_available_height)
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/layout/containers.py", line 128, in preferred_height
    dimensions = [c.preferred_height(cli, width, max_available_height) for c in self.children]
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/layout/containers.py", line 1000, in preferred_height
    cli, width - total_margin_width, max_available_height, wrap_lines),
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/layout/controls.py", line 502, in preferred_height
    content = self.create_content(cli, width, None)
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/layout/controls.py", line 614, in create_content
    ignore_case=ss.ignore_case))
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/buffer.py", line 1084, in document_for_search
    search_result = self._search(search_state, include_current_position=True)
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/buffer.py", line 1071, in _search
    result = search_once(working_index, document)
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/buffer.py", line 1061, in search_once
    document = Document(self._working_lines[i], len(self._working_lines[i]))
  File "/Users/andriimaksai/anaconda/lib/python2.7/site-packages/prompt_toolkit/document.py", line 83, in __init__
    assert isinstance(text, six.text_type), 'Got %r' % text
AssertionError: Got 'pwd'
```

**Comment by takluyver** [comment]
OK, time to call on @jonathanslenders then. It looks like a `str` is getting in where Python 2 expects `unicode`.

**Comment by jonathanslenders** [comment]
This looks like the actual issue is not the reverse search, but the fact that ``buffer._working_lines`` contains some Python2 ``str`` objects. This list is populated from the history, so I guess that the ``History`` implementation does not decode all binary objects. I can look into this, later this week or early next week.

**Comment by takluyver** [comment]
Thanks! We pre-popuate history from an SQLite database [here](https://github.com/ipython/ipython/blob/6b34d3563262aef7de63d4bde05ce9396f7e7620/IPython/terminal/interactiveshell.py#L225) - but I think the strings we get back from that should all be unicode.

**Comment by thejohnhoffer** [comment]
Is there any progress on this front? I'm considering getting involved because I'd like to be able to read ipynb in jupyter console. To do this, I'd like to load an ipynb into the SQLite history. But I'm not able to begin such a project until the history works properly.

**Comment by sboeser** [comment]
Very same error here - any progress on this?

Here is my `python -c "import IPython; print(IPython.sys_info())"`
```
{'commit_hash': u'5a894b9',
 'commit_source': 'installation',
 'default_encoding': 'UTF-8',
 'ipython_path': '/Users/sboeser/Software/VirtualEnvs/IPython/lib/python2.7/site-packages/IPython',
 'ipython_version': '5.3.0',
 'os_name': 'posix',
 'platform': 'Darwin-15.6.0-x86_64-i386-64bit',
 'sys_executable': '/Users/sboeser/Software/VirtualEnvs/IPython/bin/python',
 'sys_platform': 'darwin',
 'sys_version': '2.7.10 (default, Oct 23 2015, 19:19:21) \n[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.59.5)]'}
```

**Comment by lovelyzoo** [comment]
I am also seeing this problem on an Ubuntu 16.04 installation.

**Comment by takluyver** [comment]
I can't reproduce the error - anyone who can, can you test #10371 ?

**Comment by takluyver** [comment]
I believe this was fixed by #10371, release 5.4.
